com.onesignal.unity.ios/Runtime/iOSNotificationsManager.cs

```diff
@@ -0,0 +1,148 @@
+        //[DllImport("__Internal")] private static extern void _registerForProvisionalAuthorization(); // iOS only
+        [DllImport("__Internal")] private static extern void _notificationsAddPermissionStateChangedCallback(StateListenerDelegate callback);
+        [DllImport("__Internal")] private static extern void _notificationsSetWillShowHandler(NotificationWillShowInForegroundDelegate callback);
+        [DllImport("__Internal")] private static extern void _notificationsSetOpenedHandler(StringListenerDelegate callback);
+
+        public delegate void StateListenerDelegate(string current, string previous);
+
+        private delegate bool NotificationWillShowInForegroundDelegate(string notification);
+        private delegate void StringListenerDelegate(string response);
+        private delegate void BooleanResponseDelegate(int hashCode, bool response);
+
+        public event NotificationWillShowDelegate WillShow;
+        public event NotificationClickedDelegate Clicked;
+        public event PermissionChangedDelegate PermissionChanged;
+
+        private static iOSNotificationsManager _instance;
+
+        public iOSNotificationsManager() {
+            _instance = this;
+        }
+
+        public bool Permission {
+            get => _notificationsGetPermission();
+        }
+
+        public async Task<bool> RequestPermissionAsync(bool fallbackToSettings) {
+            var (proxy, hashCode) = WaitingProxy._setupProxy<bool>();
+            _notificationsRequestPermission(fallbackToSettings, hashCode, BooleanCallbackProxy);
+            return await proxy;
+        }
+
+        public void ClearAllNotifications() {
+            _notificationsClearAll();
+        }
+
+        public void Initialize() {
+            _notificationsAddPermissionStateChangedCallback(_onPermissionStateChanged);
+            _notificationsSetWillShowHandler(_onWillShow);
+            _notificationsSetOpenedHandler(_onOpened);
+        }
+
+        [AOT.MonoPInvokeCallback(typeof(StateListenerDelegate))] // temp
+        private static void _onPermissionStateChanged(string current, string previous) {
+            var curr = JsonUtility.FromJson<PermissionState>(current);
+            //var prev = JsonUtility.FromJson<PermissionState>(previous);
+            UnityMainThreadDispatch.Post(state => _instance.PermissionChanged?.Invoke(curr.reachable));
+        }
+
+        /// <param name="response">OSNotification</param>
+        [AOT.MonoPInvokeCallback(typeof(NotificationWillShowInForegroundDelegate))]
+        private static bool _onWillShow(string response) {
+            if (_instance.WillShow == null)
+                return true;
+
+            var notification = JsonUtility.FromJson<Notification>(response);
+            _fillNotifFromObj(ref notification, Json.Deserialize(response));
+
+            Notification resultNotif = null;
+            UnityMainThreadDispatch.Send(state => { resultNotif = _instance.WillShow(notification);});
```

Comment:
can you add a comment here that explains why we need to block the current thread using `UnityMainThreadDispatch.Send`.

Reply:
Adding in another PR